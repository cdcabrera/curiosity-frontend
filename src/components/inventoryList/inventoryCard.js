import React from 'react';
import PropTypes from 'prop-types';
import { Bullseye, Card, CardActions, CardBody, CardFooter, CardHeader, CardHeaderMain } from '@patternfly/react-core';
import { connect, reduxActions, reduxSelectors } from '../../redux';
import { MinHeight } from '../minHeight/minHeight';
import Pagination from '../pagination/pagination';
import { ToolbarFieldDisplayName } from '../toolbar/toolbarFieldDisplayName';
import { paginationHelpers } from '../pagination/paginationHelpers';
import { translate } from '../i18n/i18n';

/**
 * A hosts system inventory component.
 *
 * @fires onColumnSort
 * @fires onPage
 * @param {object} props
 * @param {boolean} props.isDisabled
 * @param {boolean} props.error
 * @param {boolean} props.fulfilled
 * @param {number} props.itemCount
 * @param {boolean} props.pending
 * @param {Function} props.t
 * @returns {Node}
 */
const InventoryList = ({
  isDisabled,
  error,
  fulfilled,
  itemCount,
  pending,
  t,
  updatedOffset,
  updatedPerPage,
  onPage,
  children
}) => {
  if (isDisabled) {
    return (
      <Card className="curiosity-inventory-card__disabled">
        <CardBody>
          <Bullseye>{t('curiosity-inventory.tab', { context: 'disabled' })}</Bullseye>
        </CardBody>
      </Card>
    );
  }

  const isLastPage = paginationHelpers.isLastPage(updatedOffset, updatedPerPage, itemCount);

  // Set an updated key to force refresh minHeight
  const minHeightContentRefreshKey =
    (fulfilled === true && itemCount < updatedPerPage && `bodyMinHeight-${updatedPerPage}-resize`) ||
    (fulfilled === true && isLastPage && `bodyMinHeight-${updatedPerPage}-resize`) ||
    (error === true && `bodyMinHeight-${updatedPerPage}-resize`) ||
    `bodyMinHeight-${updatedPerPage}`;

  return (
    <Card className="curiosity-inventory-card">
      <MinHeight key="headerMinHeight" updateOnContent>
        <CardHeader className={(error && 'hidden') || ''} aria-hidden={error || false}>
          <CardHeaderMain>
            <ToolbarFieldDisplayName />
          </CardHeaderMain>
          <CardActions className={(!itemCount && 'transparent') || ''} aria-hidden={!itemCount || false}>
            <Pagination
              isCompact
              isDisabled={pending || error}
              itemCount={itemCount}
              offset={updatedOffset}
              onPage={onPage}
              onPerPage={onPage}
              perPage={updatedPerPage}
            />
          </CardActions>
        </CardHeader>
      </MinHeight>
      <MinHeight key={minHeightContentRefreshKey} updateOnContent>
        <CardBody>
          <div className={(error && 'blur') || 'fadein'}>{children}</div>
        </CardBody>
      </MinHeight>
      <MinHeight key="footerMinHeight" updateOnContent>
        <CardFooter
          className={(error && 'hidden') || (!itemCount && 'transparent') || ''}
          aria-hidden={error || !itemCount || false}
        >
          <Pagination
            dropDirection="up"
            isDisabled={pending || error}
            itemCount={itemCount}
            offset={updatedOffset}
            onPage={onPage}
            onPerPage={onPage}
            perPage={updatedPerPage}
          />
        </CardFooter>
      </MinHeight>
    </Card>
  );
};

/**
 * Prop types.
 *
 * @type {{t: Function, listData: Array, session: object, pending: boolean, fulfilled: boolean,
 *     getHostsInventory: Function, perPageDefault: number, isDisabled: boolean, error: boolean,
 *     itemCount: number}}
 */
InventoryList.propTypes = {
  error: PropTypes.bool,
  fulfilled: PropTypes.bool,
  isDisabled: PropTypes.bool,
  itemCount: PropTypes.number,
  pending: PropTypes.bool,
  t: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{t: Function, listData: Array, session: object, pending: boolean, fulfilled: boolean,
 *     getHostsInventory: Function, perPageDefault: number, isDisabled: boolean, error: boolean,
 *     itemCount: number}}
 */
InventoryList.defaultProps = {
  error: false,
  fulfilled: false,
  isDisabled: false,
  itemCount: 0,
  pending: false,
  t: translate
};

/**
 * Create a selector from applied state, props.
 *
 * @type {Function}
 */
const makeMapStateToProps = reduxSelectors.inventoryList.makeInventoryList();

const ConnectedInventoryList = connect(makeMapStateToProps, mapDispatchToProps)(InventoryList);

export { ConnectedInventoryList as default, ConnectedInventoryList, InventoryList };
