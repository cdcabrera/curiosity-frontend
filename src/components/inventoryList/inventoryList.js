import React from 'react';
import PropTypes from 'prop-types';
import _isEqual from 'lodash/isEqual';
import { helpers } from '../../common';
import { connectTranslate, reduxActions, reduxSelectors } from '../../redux';
import Table from '../table/table';
// import { TableSkeleton } from '../table/tableSkeleton';
import { Loader } from '../loader/loader';
import GuestsList from '../guestsList/guestsList';

class InventoryList extends React.Component {
  componentDidMount() {
    this.onUpdateInventoryData();
  }

  componentDidUpdate(prevProps) {
    const { listQuery, productId } = this.props;

    if (productId !== prevProps.productId || !_isEqual(listQuery, prevProps.listQuery)) {
      this.onUpdateInventoryData();
    }
  }

  onUpdateInventoryData = () => {
    const { getHostsInventory, listQuery, isDisabled, productId } = this.props;

    if (!isDisabled && productId) {
      getHostsInventory(productId, listQuery);
    }
  };

  renderTable() {
    const { listData, listQuery, productId, t, viewId } = this.props;

    return (
      <Table
        className="curiosity-inventory-list"
        columnHeaders={[
          t('curiosity-inventory.headerName'),
          t('curiosity-inventory.headerInfrastructure'),
          t('curiosity-inventory.headerSocketsCores'),
          t('curiosity-inventory.headerLastSeen')
        ]}
        rows={listData.map(({ displayName, hardwareType, sockets, cores, lastSeen }) => ({
          cells: [displayName, hardwareType, `${sockets}/${cores}`, lastSeen],
          expandedContent: <GuestsList listQuery={listQuery} productId={productId} viewId={viewId} />
        }))}
      />
    );
  }

  render() {
    const { error, pending } = this.props;
    /*
    <TableSkeleton
            className="curiosity-inventory-list"
            columnHeaders={[
              t('curiosity-inventory.headerName'),
              t('curiosity-inventory.headerInfrastructure'),
              t('curiosity-inventory.headerSocketsCores'),
              t('curiosity-inventory.headerLastSeen')
            ]}
            cellCount={4}
            rowCount={2}
          />
     */
    return (
      <div className={`curiosity-inventory-list-wrapper fadein ${(error && 'blur') || ''}`}>
        {pending && (
          <Loader variant="table" tableProps={{ className: 'curiosity-inventory-list', colCount: 4, rowCount: 2 }} />
        )}
        {!pending && this.renderTable()}
      </div>
    );
  }
}

InventoryList.propTypes = {
  // columnHeaders: PropTypes.array,
  // rows: PropTypes.array,
  error: PropTypes.bool,
  // fulfilled: PropTypes.bool,
  getHostsInventory: PropTypes.func,
  isDisabled: PropTypes.bool,
  listData: PropTypes.array,
  pending: PropTypes.bool,
  productId: PropTypes.string.isRequired,
  listQuery: PropTypes.object.isRequired,
  t: PropTypes.func,
  viewId: PropTypes.string
};

InventoryList.defaultProps = {
  error: false,
  // fulfilled: false,
  getHostsInventory: helpers.noop,
  listData: [],
  isDisabled: helpers.UI_DISABLED_TABLE,
  pending: false,
  t: helpers.noopTranslate,
  viewId: 'inventoryList'
};

/**
 * Apply actions to props.
 *
 * @param {Function} dispatch
 * @returns {object}
 */
const mapDispatchToProps = dispatch => ({
  getHostsInventory: (id, query) => dispatch(reduxActions.rhsm.getHostsInventory(id, query))
});

/**
 * Create a selector from applied state, props.
 *
 * @type {Function}
 */
const makeMapStateToProps = reduxSelectors.inventoryList.makeInventoryList();

const ConnectedInventoryList = connectTranslate(makeMapStateToProps, mapDispatchToProps)(InventoryList);

export { ConnectedInventoryList as default, ConnectedInventoryList, InventoryList };
