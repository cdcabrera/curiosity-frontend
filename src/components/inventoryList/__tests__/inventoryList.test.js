import React from 'react';
import { shallow } from 'enzyme';
import { InventoryList } from '../inventoryList';

describe('InventoryList Component', () => {
  it('should render a non-connected component', () => {
    const props = {
      listQuery: {},
      productId: 'lorem'
    };

    const component = shallow(<InventoryList {...props} />);
    expect(component).toMatchSnapshot('non-connected');
  });
});
