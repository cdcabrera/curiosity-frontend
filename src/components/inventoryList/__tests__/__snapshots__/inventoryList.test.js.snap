// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryList Component should render a non-connected component: non-connected 1`] = `
<Table
  ariaLabel="Compact table."
  className="curiosity-inventory-list"
  columnHeaders={
    Array [
      "t(curiosity-inventory.headerName)",
      "t(curiosity-inventory.headerInfrastructure)",
      "t(curiosity-inventory.headerSocketsCores)",
      "t(curiosity-inventory.headerLastSeen)",
    ]
  }
  rows={Array []}
  summary="A generated table with one level of column headers."
  variant="compact"
/>
`;
